firestore/src/main/query_main.cc

```diff
@@ -210,29 +230,34 @@ core::Bound QueryInternal::ToBound(
         "values than were specified in the order by.");
   }
 
-  std::vector<model::FieldValue> components;
+  SharedMessage<google_firestore_v1_ArrayValue> components{{}};
+  components->values_count = CheckedSize(field_values.size());
+  components->values =
+      MakeArray<google_firestore_v1_Value>(components->values_count);
 
   for (int i = 0; i != field_values.size(); ++i) {
-    model::FieldValue field_value =
+    Message<google_firestore_v1_Value> field_value =
         user_data_converter_.ParseQueryValue(field_values[i]);
     const core::OrderBy& order_by = explicit_order_bys[i];
     if (order_by.field().IsKeyFieldPath()) {
-      components.push_back(ConvertDocumentId(field_value, internal_query));
+      components->values[i] =
+          *ConvertDocumentId(field_value, internal_query).release();
     } else {
-      components.push_back(std::move(field_value));
+      components->values[i] = *field_value.release();
     }
   }
 
-  return core::Bound{std::move(components), IsBefore(bound_pos)};
+  return core::Bound::FromValue(std::move(components), IsBefore(bound_pos));
 }
 
-model::FieldValue QueryInternal::ConvertDocumentId(
-    const model::FieldValue& from, const core::Query& internal_query) const {
-  if (from.type() != model::FieldValue::Type::String) {
+Message<google_firestore_v1_Value> QueryInternal::ConvertDocumentId(
+    const Message<google_firestore_v1_Value>& from,
+    const core::Query& internal_query) const {
+  if (GetTypeOrder(*from) != TypeOrder::kString) {
     SimpleThrowInvalidArgument(
         "Invalid query. Expected a string for the document ID.");
   }
-  const std::string& document_id = from.string_value();
+  const std::string& document_id = MakeString(from->string_value);
```

Comment:
Please make `document_id` a non-reference variable to avoid attaching a reference to a temporary.

Reply:
Done.